core/src/sumeragi/main_loop.rs

```diff
@@ -32,8 +31,12 @@ pub struct Sumeragi {
     pub max_txs_in_block: usize,
     /// Kura instance used for IO
     pub kura: Arc<Kura>,
-    /// [`iroha_p2p::Network`] actor address
```

Comment:
I agree with others.
Although making code testable requires to change the working code, it can be done with a more convinient way.
Maybe `mockall` crate might help here.

Reply:
Or maybe you can create something like:

```rust
#[cfg(test)]
pub fn set_post_procedure(procedure: SumeragiMessagePostProcedure) {}
```

And then use it in tests

Reply:
IrohaNetwork is a defined interface yes. But it only allows async api. Sync api is required for the softfork test. I will make sure closures don't leak into other crates.

Reply:
> But it only allows async api 

What do you mean? Most of the `NetworkBaseHandle` are sync expect `strart` and `wait_for_online_peers`.

Reply:
> And can you please elaborate in which way sync api is required for soft fork test?

I need to block the sumeragi thread in order to trigger the softfork at the right moment. When it calls the closure I can simply call thread sleep inside the closure.

Reply:
So have you considered `mockall` crate? I'm just not a big fan of using dyn closures here. I think there should be a better way

Reply:
we should definitely explore better ways as part of this PR. I'm not fan of `mockall` because it allows data races if you're not careful of how you're using it's API. One of the reasons we use Rust is precisely to not have data races. We can, however, replicate the compile time injection trick it uses

There is [this](https://github.com/ryo33/mry) interesting testing framework, but I haven't used it so I can't tell

Reply:
I'm afraid Mry is worse than `mockall` for reliability.